File: razer-sdk/src/test/java/com/bonepl/razersdk/animation/AnimatedFrameTest.java
package com.bonepl.razersdk.animation;

import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.color.StaticColor;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.*;

class AnimatedFrameTest {
    @Test
    void testEmptyFrameCreation() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //then
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    @Test
    void testStartWithoutFrame() {
        //when
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //then
        assertFalse(animatedFrame.hasFrame());
    }

    @Test
    void testWrongAmountOfAnimationFrames() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(0, new SimpleFrame(StaticColor.RED));
        animatedFrame.addAnimationFrame(-2, new SimpleFrame(StaticColor.RED));

        //then
        assertFalse(animatedFrame.hasFrame());
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    @Test
    void testHasFrames() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(new SimpleFrame(StaticColor.RED));
        animatedFrame.addAnimationFrame(2, new SimpleFrame(StaticColor.BLUE));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.RED);
        assertHasFrameAndMatches(animatedFrame, StaticColor.BLUE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.BLUE);
        assertFalse(animatedFrame.hasFrame());
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    @Test
    void testSingleFrameAdd() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(new SimpleFrame(StaticColor.BROWN));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.BROWN);
        assertFalse(animatedFrame.hasFrame());
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    @Test
    void testMultipleFrameAdd() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(2, new SimpleFrame(StaticColor.ORANGE));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.ORANGE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.ORANGE);
        assertFalse(animatedFrame.hasFrame());
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    @Test
    void testAnimationStop() {
        //given
        final AnimatedFrame animatedFrame = new AnimatedFrame();
        animatedFrame.addAnimationFrame(5, new SimpleFrame(StaticColor.PURPLE));
        assertHasFrameAndMatches(animatedFrame, StaticColor.PURPLE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.PURPLE);

        //when
        animatedFrame.clearFrames();

        //then
        assertFalse(animatedFrame.hasFrame());
        assertThrows(NoSuchElementException.class, animatedFrame::getFrame);
    }

    private static void assertHasFrameAndMatches(AnimatedFrame animatedFrame, Color expectedColor) {
        assertTrue(animatedFrame.hasFrame());
        assertTrue(animatedFrame.getFrame().getKeysToColors().values()
                .stream().allMatch(color -> color == expectedColor));
    }
}

File: test/razer-sdk/animation/AnimatedFrame.test.ts
import {describe, expect, test} from 'vitest';
import {AnimatedFrame} from '../../../src/razer-sdk/animation/AnimatedFrame.js';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {Color} from '../../../src/razer-sdk/color/Color.js';

function assertHasFrameAndMatches(animatedFrame: AnimatedFrame, expectedColor: Color): void {
    expect(animatedFrame.hasFrame()).toBe(true);
    const values = [...animatedFrame.pollFrame().getKeysToColors().values()];
    expect(values.every(color => color === expectedColor)).toBe(true);
}

describe('AnimatedFrameTest', () => {
    test('testEmptyFrameCreation', () => {
        //given
        const animatedFrame = new AnimatedFrame();

        //then
        expect(() => animatedFrame.pollFrame()).toThrow();
    });

    test('testStartWithoutFrame', () => {
        //when
        const animatedFrame = new AnimatedFrame();

        //then
        expect(animatedFrame.hasFrame()).toBe(false);
    });

    test('testWrongAmountOfAnimationFrames', () => {
        //given
        const animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(0, new SimpleFrame(StaticColor.RED));
        animatedFrame.addAnimationFrame(-2, new SimpleFrame(StaticColor.RED));

        //then
        expect(animatedFrame.hasFrame()).toBe(false);
        expect(() => animatedFrame.pollFrame()).toThrow();
    });

    test('testHasFrames', () => {
        //given
        const animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(new SimpleFrame(StaticColor.RED));
        animatedFrame.addAnimationFrame(2, new SimpleFrame(StaticColor.BLUE));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.RED);
        assertHasFrameAndMatches(animatedFrame, StaticColor.BLUE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.BLUE);
        expect(animatedFrame.hasFrame()).toBe(false);
        expect(() => animatedFrame.pollFrame()).toThrow();
    });

    test('testSingleFrameAdd', () => {
        //given
        const animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(new SimpleFrame(StaticColor.BROWN));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.BROWN);
        expect(animatedFrame.hasFrame()).toBe(false);
        expect(() => animatedFrame.pollFrame()).toThrow();
    });

    test('testMultipleFrameAdd', () => {
        //given
        const animatedFrame = new AnimatedFrame();

        //when
        animatedFrame.addAnimationFrame(2, new SimpleFrame(StaticColor.ORANGE));

        //then
        assertHasFrameAndMatches(animatedFrame, StaticColor.ORANGE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.ORANGE);
        expect(animatedFrame.hasFrame()).toBe(false);
        expect(() => animatedFrame.pollFrame()).toThrow();
    });

    test('testAnimationStop', () => {
        //given
        const animatedFrame = new AnimatedFrame();
        animatedFrame.addAnimationFrame(5, new SimpleFrame(StaticColor.PURPLE));
        assertHasFrameAndMatches(animatedFrame, StaticColor.PURPLE);
        assertHasFrameAndMatches(animatedFrame, StaticColor.PURPLE);

        //when
        animatedFrame.clearFrames();

        //then
        expect(animatedFrame.hasFrame()).toBe(false);
        expect(() => animatedFrame.pollFrame()).toThrow();
    });
});
